test/backends.py

```diff
@@ -0,0 +1,217 @@
+                from numbers import Number
+                if isinstance(argument, Number):
+                    return self.usr_np.abs(argument)
+                # Needed to play nicely with Jax, which frequently creates
+                # arrays of shape () when handed numbers
+                if (isinstance(argument, self.usr_np.ndarray)
+                        and argument.shape == ()):
+                    return self.usr_np.abs(argument)
+                return self.usr_np.linalg.norm(argument, normord)
+
+        return JaxPyroWrapper
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.python.gen_thermochem_code(sol=sol)
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        base_cls = module.Thermochemistry
+
+        if user_np == jnp:
+            base_cls = PythonBackend._gen_jax_wrapper(base_cls)
+
+        return base_cls(user_np)
+
+
+class CppBackend(Backend):
+    def get_name():
+        return "cpp"
+
+    def fix_arg(self, x):
+        if isinstance(x, Iterable) and not isinstance(x, str):
+            return self.pyro.VectorDouble(x)
+        return super().fix_arg(x)
+
+    def fix_result(self, result):
+        if isinstance(result, self.pyro.VectorDouble):
+            return np.array(result)
+        return super().fix_result(result)
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.cpp.gen_thermochem_code(sol=sol, namespace=name)
+
+
+class FortranBackend(Backend):
+    def get_name():
+        return "fortran90"
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        return getattr(module, f"libpyro_fortran90_{mechname}")
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.fortran90.gen_thermochem_code(
+            sol=sol, module_name=name
+        )
+
+
+BACKENDS: typing.Dict[str, Backend] = {
+    CppBackend.get_name(): CppBackend,
+    FortranBackend.get_name(): FortranBackend,
+    PythonBackend.get_name(): PythonBackend
+}
+
+
+class Pyrometheus:
+    def __init__(self, mechname: str, backend: Backend, user_np):
+        self.backend = backend
+        self.pyro = self.backend.extract_interface(
+            importlib.import_module(f"libpyro_{self.backend.get_name()}_{mechname}"),
+            mechname,
+            user_np
+        )
+
+    def __getattr__(self, name: str):
+        attrib = getattr(self.pyro, name)
+        if not callable(attrib):
+            return attrib
+
+        def wrapper(*args, **kwargs):
+            assert (len(kwargs) == 0)
+
+            backend_instance = self.backend(self.pyro)
+
+            return backend_instance.fix_result(
+                attrib(*[backend_instance.fix_arg(x) for x in args])
+            )
+
+        return wrapper
+
+
+def pyro_init(mechname: str, user_np, request) -> (ct.Solution, Pyrometheus):
```

Comment:
I think `pyro_init` is somewhat vague here. Can we think of something more descriptive? Also, please add a docstring for this method.

Reply:
`create_pyro_[instance/representation]_[for/from]_mech`, or simply `create_pyro_representation`?

Reply:
Thoughts on this @henryleberre?

Reply:
I'll resolve and change afterwards.